### PR TITLE
Components and Codecs

### PR DESCRIPTION
**Pull request by supersaiyansubtlety**
Mostly completes `n/m/components`.
Completes `Codecs` and `n/m/network/codec`, along with some other codec classes.
Also some misc. things.

### CODE REVIEW

**Review by OroArmor — APPROVED** [review]

File: mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping [hunk 1]
```diff
@@ -7,6 +7,11 @@ CLASS net/minecraft/unmapped/C_azfkqhlm net/minecraft/component/type/AttributeMo
 	METHOD m_addoqghi compute (DLnet/minecraft/unmapped/C_yuycoehb;)D
 		ARG 1 base
 		ARG 3 slot
+	METHOD m_alrhqkhd (Ljava/util/List;)Lnet/minecraft/unmapped/C_azfkqhlm;
+		ARG 0 modifiers
+	METHOD m_bpimhffk forEachModifierOfSlots (Lnet/minecraft/unmapped/C_qfoqahef;Ljava/util/function/BiConsumer;)V
```
Comment (eristhea) [inline mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping]:
```suggestion
	METHOD m_bpimhffk forEachModifier (Lnet/minecraft/unmapped/C_qfoqahef;Ljava/util/function/BiConsumer;)V
```
the extra words seem unnecessary and make the name awkward

File: mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping [hunk 2]
```diff
@@ -19,7 +24,14 @@ CLASS net/minecraft/unmapped/C_azfkqhlm net/minecraft/component/type/AttributeMo
 	METHOD m_pxeavngw builder ()Lnet/minecraft/unmapped/C_azfkqhlm$C_vinsjuvn;
 	METHOD m_wfqjjama (Ljava/text/DecimalFormat;)V
 		ARG 0 format
+	METHOD m_zgatrfmf forEachModifierOfSlot (Lnet/minecraft/unmapped/C_yuycoehb;Ljava/util/function/BiConsumer;)V
```
Comment (eristhea) [inline mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping]:
```suggestion
	METHOD m_zgatrfmf forEachModifier (Lnet/minecraft/unmapped/C_yuycoehb;Ljava/util/function/BiConsumer;)V
```

File: mappings/net/minecraft/component/type/CustomModelDataComponent.mapping [hunk 3]
```diff
@@ -1 +1,9 @@
 CLASS net/minecraft/unmapped/C_qtexserh net/minecraft/component/type/CustomModelDataComponent
+	FIELD f_ampyzsae EMPTY Lnet/minecraft/unmapped/C_qtexserh;
+	METHOD m_alzguglw getString (I)Ljava/lang/String;
+	METHOD m_ggenqlfs getOrNull (Ljava/util/List;I)Ljava/lang/Object;
+		ARG 0 list
+		ARG 1 i
```
Comment (eristhea) [inline mappings/net/minecraft/component/type/CustomModelDataComponent.mapping]:
```suggestion
		ARG 1 index
```

File: mappings/net/minecraft/util/dynamic/Codecs.mapping [hunk 4]
```diff
@@ -72,36 +129,154 @@ CLASS net/minecraft/unmapped/C_tyjpezxh net/minecraft/util/dynamic/Codecs
...
+	METHOD m_shkjkhku (Lcom/mojang/authlib/properties/Property;)Ljava/util/Optional;
+		ARG 0 property
+	METHOD m_sltnwzrc createExclusiveMinRangedFloat (FFLjava/util/function/Function;)Lcom/mojang/serialization/Codec;
+		ARG 0 min
+		ARG 1 max
+		ARG 2 messageFactory
+	METHOD m_szibgwdf (Lcom/mojang/datafixers/util/Either;)Lcom/mojang/authlib/properties/PropertyMap;
+		ARG 0 stringMapOrProperties
+	METHOD m_tmfqobpn (Ljava/util/function/BiFunction;Lcom/mojang/datafixers/util/Either;)Lcom/mojang/serialization/DataResult;
+		ARG 1 pOrI
```
Comment (eristhea) [inline mappings/net/minecraft/util/dynamic/Codecs.mapping]: this is a lambda but still. what does this *mean*

**Review by eristhea — APPROVED** [review]
looks great, just a couple nitpicks!

File: mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping (hunk 1 above)
Comment (supersaiyansubtlety) [reply mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping]: I think it makes it clearer that it uses the passed slots to filter which attributes the passed function is applied to.
It is a bit awkward though.

Do you think it's clear enough without the suffix?

Comment (eristhea) [reply mappings/net/minecraft/component/type/AttributeModifiersComponent.mapping]: i think it's definitely enough, the parameters are context

**Pull request merged**